### ISSUE TITLE
I want use custom vpn server with Blokada

### ISSUE THREAD

**Issue by thecyberd3m0n** [blokadaorg/blokada#520]
I'm often using personal vpn to access computers on LAN. But I can't use it with Blokada. Blokada supports vpn, but only their servers.
Allow user to add custom server as vpn to Blokada.

### Steps to reproduce
1. Try to use custom vpn with Blokada

### Expected behaviour

App has option to add custom vpn

### Actual behaviour

App uses hardcoded vpn server
### Configuration
1. Blokada version

2. Device model

3. Android version / ROM

### Questions
1. How long have you been using Blokada? 

2. How long time ago has the problem started to happen? 

3.  How often does this problem happen to you?

**Comment by kar** [comment]
Relevant #111 

We do plan to bring support to Wireguard and OpenVPN protocols, this will solve your request. It's a bit of work though. Anyone interested in this please vote on this ticket so that we can judge the interest.

**Comment by Perelandra0x309** [comment]
+1 for Blokada with Wireguard

**Comment by 14553** [comment]
This is Not a issue any more in Stable 4

**Comment by soredake** [comment]
Any progress for 5 version?

**Comment by L31FY** [comment]
> > This is Not a issue any more in Stable 4
> 
> Can you elaborate? I cannot see how I'm able to set custom vpn servers in v4, just choose between Blokada provided ones
> 
> This would be much appreciated, if lack of this feature is due to monetization, I wouldn't mind paying one time fee to unlock this and support the devs.
> 
> I care about self hosted wireguard for security and in order to provide access to internal services, implement my own packet montioring etc.
> 
> Thanks

It's a lack of having someone to do the dev work. It's not as easy as it sounds. It's quite intense actually. They keep making the Android base more complicated and restricted with networking all the time too and this complicates things. It's like basically building a whole other app into this one by doing this, from the ground up. It's not the money so much the time and work to build the code then actually make it work.

**Comment by leshy** [comment]
> It's a lack of having someone to do the dev work. It's not as easy as it sounds. It's quite intense actually. They keep making the Android base more complicated and restricted with networking all the time too and this complicates things. It's like basically building a whole other app into this one by doing this, from the ground up. It's not the money so much the time and work to build the code then actually make it function properly.

Yes that's true, this whole conversation is due to android offering only single network processing layer in the form of a "VPN", you shouldn't need to build a fully functional VPN client if you are doing ad blocking..

I also realized that if someone is deploying their own wireguard, they might as well implement DIY blocking on the server, using the same lists and strategy you use. I can drop an update here for interested parties if I go in this direction.

Ok, thanks for the clarification and amazing work! love the app, I know you have the donation button, I'd subscribe if you had a patreon as well.

**Comment by maffe** [comment]
> I also realized that if someone is deploying their own wireguard, they might as well implement DIY blocking on the server, using the same lists and strategy you use. I can drop an update here for interested parties if I go in this direction.

I considered this, but that way it’s not possible to define app-specific exceptions.

**Comment by Finoway** [comment]
Hello, I'm very interested personally and it would be very useful to me. Do you think it will be added ?

**Comment by itcat4u43** [comment]
Any updates to this issue?

**Comment by gelavat** [comment]
Isn't it possible to just add an option to forward the traffic to a second installed VPN on Android or iOS?
This way, firstly it is much less work for you, and also it would allow us to use any VPN app on top of Blokada, so we would have all benefits of the other app too, added to the benefits of Blokada.

It would be like two nested VPNs, traffic goes first to Blokada, blocks whatever it needs to block, and then forwards what has been seens as valid to the other VPN of our choice. It seems that VPNs are like listed in Android, so from Blokada there should be a choice on which VPN we want to forward the traffic.
